### PR TITLE
Revert IO<E, A> back to IO<A>

### LINKED ISSUE #175: ["Request"] Revert `IO<E, A>` back to `IO<A>`
`IO<E, A>` is will be removed before release given the improved [`suspend` encoding](https://github.com/arrow-kt/arrow-fx/pull/169) moving forward for Arrow Fx in Kotlin.

For that reason `IO<E, A>` needs to be reverted back to its original `IO<A>`.

### PR DESCRIPTION
**Pull request by aballano**
This reverts commit 0d7c9c2b58703e380b222427581c15f74b3db214.

Fixes #175 

◀️  Depends on https://github.com/arrow-kt/arrow-ank/pull/37
▶️ Unlocks https://github.com/arrow-kt/arrow-ank/pull/41

### CODE REVIEW

File: arrow-fx/src/main/kotlin/arrow/fx/IO.kt [hunk 1]
```diff
@@ -1065,7 +1187,7 @@ fun <E, A, B, E2 : E> IOOf<E, A>.flatMap(f: (A) -> IOOf<E2, B>): IO<E2, B> =
  * }
  * ```
  */
-fun <E, A> IO<E, EitherOf<E, A>>.flattenEither(): IO<E, A> =
+fun <E : Throwable, A> IO<EitherOf<E, A>>.flattenEither(): IO<A> =
```
Comment (aballano) [inline arrow-fx/src/main/kotlin/arrow/fx/IO.kt:1190]: 👀 @nomisRev @JorgeCastilloPrz Although not as useful as before, I left this functions here but restraining E to throwable, so they can still be used in certain situations

File: arrow-fx/src/main/kotlin/arrow/fx/IOConnection.kt [hunk 2]
```diff
@@ -1,111 +1,46 @@
...
+fun IOConnection.toDisposable(): Disposable = { cancel().fix().unsafeRunSync() }
+typealias IOConnection = KindConnection<ForIO>
 
 internal class IOContext(val connection: IOConnection) : AbstractCoroutineContextElement(IOContext) {
   companion object Key : CoroutineContext.Key<IOContext>
 }
 
-internal sealed class IOConnection {
+@Suppress("UNUSED_PARAMETER", "FunctionName")
+fun IOConnection(dummy: Unit = Unit): IOConnection = KindConnection(MD) { it.fix().unsafeRunAsync { } }
```
Comment (nomisRev) [inline arrow-fx/src/main/kotlin/arrow/fx/IOConnection.kt:19]: Should be replaced with `IOConnection` from `0.11.0`.

File: arrow-fx/src/main/kotlin/arrow/fx/IOFrame.kt [hunk 3]
```diff
@@ -14,53 +14,40 @@ import arrow.fx.IO.Pure
...
+  override operator fun invoke(a: A): R
 
-  companion object {
+  fun recover(e: Throwable): R
 
-    internal class Redeem<E, A, B>(val ft: (Throwable) -> B, val fe: (E) -> B, val fb: (A) -> B) : IOFrame<E, A, IO<Nothing, B>> {
-      override fun invoke(a: A): IO<Nothing, B> = Pure(fb(a))
-      override fun recover(e: Throwable): IO<Nothing, B> = Pure(ft(e))
-      override fun handleError(e: E): IO<Nothing, B> = Pure(fe(e))
+  fun fold(value: Either<Throwable, A>): R =
```
Comment (nomisRev) [inline arrow-fx/src/main/kotlin/arrow/fx/IOFrame.kt:22]: Where is this used?

File: arrow-fx/src/main/kotlin/arrow/fx/KindConnection.kt [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+ * Connection for kinded type [F].
+ *
+ * A connection is represented by a composite of [cancel] functions,
+ * [cancel] is idempotent and all methods are thread-safe & atomic.
+ *
+ * The cancellation functions are maintained in a stack and executed in a FIFO order.
+ */
+@Deprecated(message = "Cancelling operations through KindConnection will not be supported anymore." +
+  "In case you need to cancel multiple processes can do so by using cancellable and composing cancel operations using parMapN")
+sealed class KindConnection<F> {
```
Comment (nomisRev) [inline arrow-fx/src/main/kotlin/arrow/fx/KindConnection.kt:30]: Removed in `0.10.5`

File: arrow-fx/src/test/kotlin/arrow/fx/SemaphoreTest.kt [hunk 5]
```diff
@@ -135,10 +135,10 @@ class SemaphoreTest : ArrowFxSpec() {
     }
 
     tests("UncancellableSemaphore") { Semaphore.uncancellable(it, IO.async()) }
-    tests("CancellableSemaphore") { Semaphore(it, IO.concurrent()) }
+    tests("cancellableSemaphore") { Semaphore(it, IO.concurrent()) }
 
-    "CancellableSemaphore - supports cancellation of acquire" {
-      Semaphore(0, IO.concurrent<Nothing>()).flatMap { s ->
+    "cancellableSemaphore - supports cancellation of acquire" {
```
Comment (nomisRev) [inline arrow-fx/src/test/kotlin/arrow/fx/SemaphoreTest.kt:140]: Why these changes?

File: arrow-fx/src/test/kotlin/arrow/fx/MVarTest.kt [hunk 6]
```diff
@@ -323,27 +324,28 @@ class MVarTest : ArrowFxSpec() {
...
+          val fallback = sleep(200.milliseconds).followedBy(IO.just(0))
+          val res = !IO.raceN(finished.get(), fallback)
+          !effect { res shouldBe Right(0) }
+        }.shouldBeEq(IO.unit, IO.eq())
       }
     }
 
     tests("UncancellableMVar", MVar.factoryUncancellable(IO.async()))
-    concurrentTests("CancellableMVar", MVar.factoryCancellable(IO.concurrent()))
+    concurrentTests("cancellableMVar", MVar.factoryCancellable(IO.concurrent()))
```
Comment (nomisRev) [inline arrow-fx/src/test/kotlin/arrow/fx/MVarTest.kt:346]: ?

File: arrow-fx/src/test/kotlin/arrow/fx/SemaphoreTest.kt (hunk 5 above)
Comment (aballano) [reply arrow-fx/src/test/kotlin/arrow/fx/SemaphoreTest.kt:140]: Because https://github.com/arrow-kt/arrow-fx/pull/130/files#diff-408b5ccd8b3587c3160fb876cf3b731bR141

File: arrow-fx-kotlinx-coroutines/src/main/kotlin/arrow/integrations/kotlinx/extensions.kt [hunk 7]
```diff
@@ -28,35 +24,24 @@ import kotlin.coroutines.resumeWithException
...
  * This will make sure that the source [IO] is cancelled whenever it's [CoroutineScope] is cancelled.
  */
-suspend fun <A> IOOf<Nothing, A>.suspendCancellable(): A =
+suspend fun <A> IOOf<A>.suspendCancellable(): A =
   suspendCancellableCoroutine { cont ->
     if (cont.isActive) {
-      val disposable = fix().unsafeRunAsyncCancellable { either ->
-        either.fold(cont::resumeWithException) { cont.resume(it) }
+      val disposable = fix().unsafeRunAsyncCancellable { result ->
+        result.fold<Unit>(cont::resumeWithException, cont::resume)
```
[the remark this answers is not in the document]
Comment (aballano) [reply arrow-fx-kotlinx-coroutines/src/main/kotlin/arrow/integrations/kotlinx/extensions.kt:31]: No idea, just removed the explicit type to see if it compiles

File: arrow-fx/src/test/kotlin/arrow/fx/SemaphoreTest.kt (hunk 5 above)
Comment (nomisRev) [reply arrow-fx/src/test/kotlin/arrow/fx/SemaphoreTest.kt]: Ah, regression on correcting style 😂

File: arrow-fx/src/main/kotlin/arrow/fx/internal/IOBracket.kt [hunk 8]
```diff
@@ -87,86 +95,62 @@ internal object IOBracket {
...
+    val cancel: CancelToken<ForIO> = applyRelease(ExitCase.Cancelled).fix().uncancellable()
 
     // Unregistering cancel token, otherwise we can have a memory leak;
     // N.B. conn.pop() happens after the evaluation of `release`, because
     // otherwise we might have a conflict with the auto-cancellation logic
-    override fun recover(e: Throwable): IO<E, B> =
-      IO.Bind(IO.ContextSwitch(applyRelease(ExitCase2.Exception(e)), IO.ContextSwitch.makeUncancellable, disableUncancellableAndPop),
-        ReleaseRecoverException(e))
+    override fun recover(e: Throwable): IO<B> =
+      IO.Bind(
```
Comment (aballano) [inline arrow-fx/src/main/kotlin/arrow/fx/internal/IOBracket.kt:130]: @nomisRev I brought back this as well FYI

Comment (nomisRev) [reply arrow-fx/src/main/kotlin/arrow/fx/internal/IOBracket.kt:130]: 👌

File: arrow-fx-kotlinx-coroutines/src/main/kotlin/arrow/integrations/kotlinx/extensions.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (aballano) [reply arrow-fx-kotlinx-coroutines/src/main/kotlin/arrow/integrations/kotlinx/extensions.kt:31]: And it doesn't, need to add it explicitely 😂

[the remark this answers is not in the document]
Comment (nomisRev) [reply arrow-fx-kotlinx-coroutines/src/main/kotlin/arrow/integrations/kotlinx/extensions.kt:31]: Cool. Didn't know you could fix bypassing `<Unit>` to `fold`.
I've always made anonymous lamdas
